stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4182 @@
+//! downloader never touches disk; it only needs to do so when it is considering new sortitions and
+//! new chain tips.
+
+use std::collections::{BTreeMap, HashMap, HashSet, VecDeque};
+use std::convert::TryFrom;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::io::{Read, Write};
+use std::net::{IpAddr, SocketAddr};
+
+use rand::seq::SliceRandom;
+use rand::{thread_rng, RngCore};
+use stacks_common::types::chainstate::{
+    BlockHeaderHash, ConsensusHash, PoxId, SortitionId, StacksBlockId,
+};
+use stacks_common::types::net::{PeerAddress, PeerHost};
+use stacks_common::types::StacksEpochId;
+use stacks_common::util::hash::to_hex;
+use stacks_common::util::secp256k1::{Secp256k1PrivateKey, Secp256k1PublicKey};
+use stacks_common::util::{get_epoch_time_ms, get_epoch_time_secs, log};
+use wsts::curve::point::Point;
+
+use crate::burnchains::{Burnchain, BurnchainView, PoxConstants};
+use crate::chainstate::burn::db::sortdb::{
+    BlockHeaderCache, SortitionDB, SortitionDBConn, SortitionHandleConn,
+};
+use crate::chainstate::burn::BlockSnapshot;
+use crate::chainstate::nakamoto::{
+    NakamotoBlock, NakamotoBlockHeader, NakamotoChainState, NakamotoStagingBlocksConnRef,
+};
+use crate::chainstate::stacks::db::StacksChainState;
+use crate::chainstate::stacks::{
+    Error as chainstate_error, StacksBlockHeader, TenureChangePayload,
+};
+use crate::core::{
+    EMPTY_MICROBLOCK_PARENT_HASH, FIRST_BURNCHAIN_CONSENSUS_HASH, FIRST_STACKS_BLOCK_HASH,
+};
+use crate::net::api::gettenureinfo::RPCGetTenureInfo;
+use crate::net::chat::ConversationP2P;
+use crate::net::db::{LocalPeer, PeerDB};
+use crate::net::http::HttpRequestContents;
+use crate::net::httpcore::{StacksHttpRequest, StacksHttpResponse};
+use crate::net::inv::epoch2x::InvState;
+use crate::net::inv::nakamoto::{NakamotoInvStateMachine, NakamotoTenureInv};
+use crate::net::neighbors::rpc::NeighborRPC;
+use crate::net::neighbors::NeighborComms;
+use crate::net::p2p::PeerNetwork;
+use crate::net::server::HttpPeer;
+use crate::net::{Error as NetError, Neighbor, NeighborAddress, NeighborKey};
+use crate::util_lib::db::{DBConn, Error as DBError};
+
+/// Download states for an historic tenure.  This is a tenure for which we know the hashes of the
+/// start and end block.  This includes all tenures except for the two most recent ones.
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum NakamotoTenureDownloadState {
+    /// Getting the tenure-start block
+    GetTenureStartBlock(StacksBlockId),
+    /// Waiting for the child tenure's tenure-start block to arrive, which is usually (but not
+    /// always) handled by the execution of another NakamotoTenureDownloader.  The only
+    /// exceptions are as follows:
+    ///
+    /// * if this tenure contains the anchor block, and it's the last tenure in the
+    /// reward cycle.  In this case, the end-block must be directly fetched, since there will be no
+    /// follow-on NakamotTenureDownloader in the same reward cycle who can provide this.
```

Comment:
If this tenure happens to be both the last tenure in the reward cycle and the tenure which contains the anchor block, then if it were to enter the `WaitForTenureEndBlock` state, it would get stuck forever (at least, this was the case before the timeout logic was added).  This is because the downloader, when in this state, is expecting the `NakamotoTenureDownloaderSet` to provide it with _another_ `NakamotoTenureDownloader`'s tenure-start block in order to continue.  This mode of operation saves us from downloading the same tenure-start blocks twice.

However, this condition can never arise in this specific case, because _there is no subsequent tenure that can be fetched_.  This is because the aggregate public key for the next tenure (which lies in the next reward cycle) hasn't been processed yet, because the tenure which commits to it (the one waiting for the tenure-end block) hasn't even been downloaded yet.  Hence this exception.

Reply:
If this tenure happens to be both the last tenure in the reward cycle and the tenure which contains the anchor block, then if it were to enter the `WaitForTenureEndBlock` state, it would get stuck forever (at least, this was the case before the timeout logic was added).  This is because the downloader, when in this state, is expecting the `NakamotoTenureDownloaderSet` to provide it with _another_ `NakamotoTenureDownloader`'s tenure-start block in order to continue.  This mode of operation saves us from downloading the same tenure-start blocks twice.

However, this condition can never arise in this specific case, because _there is no subsequent tenure that can be fetched_.  This is because the aggregate public key for the next tenure (which lies in the next reward cycle) hasn't been processed yet, because the tenure which commits to it (the one waiting for the tenure-end block) hasn't even been downloaded yet.  Hence this exception.